sound/soc/sof/control.c

```diff
@@ -65,6 +65,37 @@ static inline u32 ipc_to_mixer(u32 value, u32 *volume_map, int size)
 	return i - 1;
 }
 
+static void snd_sof_refresh_control(struct snd_sof_control *scontrol)
+{
+	struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
+	struct snd_soc_component *scomp = scontrol->scomp;
+	enum sof_ipc_ctrl_type ctrl_type;
+	int ret;
+
+	if (!scontrol->updated_in_firmware)
+		return;
+
+	if (!pm_runtime_active(scomp->dev))
+		return;
+
+	if (scontrol->cmd == SOF_CTRL_CMD_BINARY)
+		ctrl_type = SOF_IPC_COMP_GET_DATA;
+	else
+		ctrl_type = SOF_IPC_COMP_GET_VALUE;
+
+	/* set the ABI header values */
+	cdata->data->magic = SOF_ABI_MAGIC;
+	cdata->data->abi = SOF_ABI_VERSION;
+
+	/* refresh the component data from DSP */
+	scontrol->updated_in_firmware = false;
```

Comment:
shouldnt this be updated only if the ipc below succeeds?

Reply:
Good point, let me think about this.
The reason why it is moved here is that if we clear it after the IPC then we would have opened a small window for a race.
On the other hand if the IPC fails then either the firmware is dead or for some reason the we can not read the control's value. In the first case there is not much point of retrying, on the second one how likely the second similar command would succeed?

Reply:
I think setting back `updated_in_firmware` to `true` in case of IPC failure should do it without introducing race, but I'm not convinced that it will be beneficial.
User space gets notification, it tries to read the data, we got IPC error, the value is not updated, user space takes the old data as new one and it is happy, not requesting it again.